### DISCUSSION TITLE
Primi risultati con Amazon Transcribe

### DISCUSSION THREAD

**Discussion by pdonadeo** [pdonadeo/e-uscito-joypad#9] [General]
Ho fatto passare l'episodio di ieri (48). Allego l'output in formato JSON e SRT (sottotitoli):
1. [ep48_json.txt](https://github.com/pdonadeo/e-uscito-joypad/files/8598693/ep48_json.txt)
2. [ep48_srt.txt](https://github.com/pdonadeo/e-uscito-joypad/files/8598694/ep48_srt.txt)

Poi ho scritto quattro righe di Python in croce:

```python
import json

def decode_speaker(label):
    if label == "spk_0":
        return "Fossa"
    elif label == "spk_1":
        return "Zampa"
    elif label == "spk_2":
        return "Bordone"
    else:
        return label

ep48 = json.loads(open('con_alternative/ep48.json', 'r').read())['results']
segments = list(zip(ep48['speaker_labels']['segments'], ep48['segments']))

for seg in segments:
    start_time = seg[0]['start_time']
    end_time = seg[0]['end_time']
    speaker_label = seg[0]['speaker_label']
    speaker = decode_speaker(speaker_label)
    try:
        text = seg[1]['alternatives'][1]['transcript']
    except IndexError:
        text = seg[1]['alternatives'][0]['transcript']
    
    print(f"""{speaker} ha detto (da {start_time}s a {end_time}s): "{text}"\n""")
```

E ho ottenuto un file di testo fatto così: [ep48.txt](https://github.com/pdonadeo/e-uscito-joypad/files/8598699/ep48.txt).

Ci vedo comunque un sacco di lavoro da fare… Vanno comunque riletti tutti, per estrarre un file statico con i dati che ci servono. Oppure potremmo buttare tutto dentro un [DB e fare delle ricerche full text](https://www.postgresql.org/docs/14/textsearch.html). Devo fare un po' di esperimenti.

Intanto avete qualcosa a cui dare un'occhiata.

**Comment by pdonadeo** [comment]
Dopo un ulteriore post processing sono riuscito ad ottenere questa cosa: [postprocess.txt](https://github.com/pdonadeo/e-uscito-joypad/files/8599126/postprocess.txt). Si tratta di un file HTML ma GitHub non me lo fa allegare, quindi rinominatelo.

È il transcript dell'episodio 48, con in rosso le parole che hanno un basso livello di affidabilità secondo gli algoritmi di AWS. Ed in effetti molte sono sbagliate.

**Reply by bionicco** [reply]
Al di là dei miei dubbi sull'utilità della trascrizione...la versione html può diventare facilmente un'interfaccia per la creazione dei contenuti.
Per esempio basterebbe in ogni riga inserire un pulsante che permetta di aggiungere dei metadati a quella frase per poterli catalogare (anche solo generando un json lato client da inviare nel metodo di raccolta dati)

**Reply by pdonadeo** [reply]
Assolutamente d'accordo! Ad ogni modo accantonerei per un attimo questa "cosa della trascrizione" che come ho scritto nell'altro thread mi serviva solo per capire "se si può fare o non ha nessun senso".

Dovremmo pensare alla funzionalità, partendo dalla UX dell'utente.

L'idea inizialmente è stata tua: tu come ti immagini questa funzione di ricerca, operativamente? Cosa scrivi? Fa autocompletamento? Di cosa? Quando premi invio cosa ottieni?

Insomma facciamoci queste domande, secondo me.

**Reply by bionicco** [reply]
Proviamo a buttare giù qualche idea.

In generale metterei una lista degli episodi, ognuno con la sua card con varie informazioni (eventualmente nascoste da un accordion o simile): titolo, immagine, descrizione, links e lista dei contenuti (giochi, frasi o argomenti, con relativo minutaggio)
E su questi un input di ricerca che durante la scrittura filtra subito le puntate che contengono quell'input in uno dei campi (lasciandole nell'ordine in cui sono già). Eventualmente potrebbero comparire dei suggerimenti legati all'input presi dai vari contenuti delle puntate, che se selezionati cambiano l'input attuale (e di conseguenza il filtraggio). Premere invio fa soltanto sparire la lista dei suggerimenti e toglie il focus dall'input
Questa lista la metterei sotto la parte che dice "si è uscito",  "no non è uscito", con in alto il campo di ricerca (che scorrendo in basso i risultati si va a piazzare tipo header il alto)

Eventualmente possiamo fare una seconda lista (da capire come accedervi, potrebbe essere insieme alla precedente in un tab) con una scheda per ogni videogame (e qui invece della lista dei contenuti ci sarebbe la lista degli episodi in cui viene nominato) ovviamente ordinati in base al numero di citazioni. Il meccanismo di ricerca potrebbe essere lo stesso...o addirittura l'input di ricerca essere lo stesso che filtra entrambe le liste

**Reply by MarcoJul** [reply]
Ciao a tutti, mi unisco anche io alla discussione, sperando di poter dare qualche contributo! 

Tenendo considerate entrambe le proposte (motore di ricerca e "frasi d'autore") provo a esprimere qualche considerazione. Parlo per ora lato frontend, provo a immaginare una struttura di pagina con le sue varie funzioni:

**1. Frasi celebri:** l'idea mi piace, la immaginerei come una sorta di database/json da cui prendere una frase a ogni refresh e farla comparire in homepage stile "citazione di tale rivista" o le varie frasi inserite sulle custodie dei giochi, in maniera più o meno parodistica (da capire chi decide le "frasi celebri"):

es:
> _"É una palla al cazzo"_
> Gioco ambientato a Tokyo - Nome dell'autore 

anche se detta così secondo me volano le denunce, forse starei sulle considerazioni di massima o le varie battute sulle batterie ricaricabili del controller Xbox.

**2. Motore di ricerca:** immagino l'utilizzo-tipo che un utente necessiti (mi ci sono pure riscontrato qualche volta, senza elaborarne l'idea e proporla come fatto da @bionicco ): sento la _punt-esodio_, e ascolto quel che si dice; dopo vario tempo mi capita sottomano il gioco e vado a risentire come se ne parlava. Il concetto vale anche se inizio ad ascoltare ora Joypad, non recupero le puntate precedenti (male:) ), mi trovo fra le mani il gioco e provo a cercarlo per capire se ne hanno parlato. Mi aspetto di trovare quindi la/le puntata/e interessata/e, con il timestamp dell'argomento.

Per l'impostazione della pagina, come già suggerito, l'idea che mi sembra più fattibile è una lista-accordion con l'elenco delle puntate (in alto la più recente, all'apertura della tab viene inserita la scaletta della puntata e/o argomenti inerenti).

Sull'**elenco titoli**, immagino potremmo già inserire in hard code Death Stranding come titolo più citato:D 
A parte gli scherzi, se si vuole mantenere l'ordine di questa lista per numero di citazioni credo serva già impostare il relativo database in tal senso, in modo che a ogni puntata nuova non impazziamo a inserire le varie occurrency.

Al momento sto cercando per quanto possibile di non entrare troppo nel particolare, ma di partire da un'idea complessiva e poi andare a lavorare sui dettagli.

Grazie intanto a tutti per il lavoro fatto!

**Reply by SKApretto71** [reply]
Faccio il guastafeste... ma con [Azure Speech Services](https://docs.microsoft.com/en-us/azure/cognitive-services/speech-service/speech-to-text) avete fatto una prova?

**Comment by pdonadeo** [comment]
@bionicco: hai descritto la funzionalità e mi piace in linea di massima, c'è solo da capire come disegnare questa cosa perché la lista che hai descritto, che si filtra via via che uno scrive, su mobile la vedo davvero malissimo:

>  titolo, immagine, descrizione, links e lista dei contenuti (giochi, frasi o argomenti, con relativo minutaggio)

48 card come questa esce un paginone senza senso. Insomma: ho capito cosa vorresti, sono d'accordo, ma c'è bisogno di una UX decente. Domani coinvolgo la mia collega grafica ed esperta di UX, vediamo se alla pausa caffè le strappo qualche suggerimento. Tra l'altro: secondo voi ho fatto io il CSS? Se l'avessi fatto io sarebbe uscito [qualcosa di orripilante](https://www.theworldsworstwebsiteever.com/).

@MarcoJul: frasi celebri... io onestamente per ora lascerei stare questa cosa nel senso che è interessante ma davvero non saprei come collocarla e soprattutto come realizzarla. Voglio dire: che cos'è una frase celebre? Una battuta che mi fa morire magari a te non dice niente. Mi concentrerei sull'idea del motore di ricerca perché:
1. è più semplice capire di cosa stiamo parlando
2. richiede solo risorse statiche o giù di lì, non serve un database
3. il mio tempo è limitato, dedico a questo progetto la sera, non tutte le sere

Detto questo non voglio accantonare l'idea delle frasi celebri, mi sembra molto carina: dico solo di tenerla come "seconda cosa da sviluppare", mentre nel frattempo ci dedichiamo alla prima e magari capiamo anche meglio cosa sono le frasi celebri.

@SKApretto71: no, non ho provato Azure ma AWS per il semplice e banale motivo che conosco AWS per lavoro e ho un account, motivo per cui fare una prova mi è costato 10 minuti. Tu hai esperienza diretta di Azure Speech Services?

**Reply by bionicco** [reply]
Su come fare l'interfaccia sicuramente c'è da studiare un po' e magari fare qualche prova. Per esempio l'episodio può essere un elemento più piccolo con solo alcuni dati (card o riga di una tabella, etc) e i maggiori dettagli (contenuti, link, descrizione, etc....) potrebbero stare in una modale o addirittura una pagina.
Posso provare nei prossimi giorni a fare una bozza di interfaccia con dati mockati e vediamo come risulta.
Sicuramente un aiuto per il css (o anche solo un bozzetto) non sarebbe male, perchè anche io non sono esperto nel disegnare UX e UI.

In ogni caso credo che dovremmo decidere come vogliamo raccogliere i dati, perchè mi pare che ci siano due groddi filoni in ballo:
A) Fare la trascrizione degli episodi e su questo avere la possibilità di fare una ricerca testuale.
Come  vantaggio c'è sicuramente un minor bisogno di manovalanza e la possibilità di avere accesso a tutto quello che è stato detto, però le informazioni sarebbero "piatte" (a meno di non applicare ulteriori livelli di elaborazione) perchè un titolo semplicemente nominato darebbe lo stesso risultato di quel titolo presente in una recensione. Inoltre forse sarebbe inevitabile una parte di servizi e db
B) Ricercare manualmente negli episodi le informazioni. Un gran lavoro nel riascoltare tutto (ma divisibile fra tutti i volontari) ma la possibilità di avere delle informazioni più arricchite
Io preferirei la B, ma parliamone.
Se optassimo per questa scelta il primo passo sarebbe quello di decidere il formato dei dati (con relativa indipendenza dall'interfaccia che possiamo elaborare successivamente) così da iniziare subito con la parte di raccolta dati (creando un form), che sicuramente porta via molto tempo, e che può essere fatta in parallelo con la creazione e programmazione dell'interfaccia

**Reply by MarcoJul** [reply]
Sul lato Frontend, che sia tramutare la grafica proposta dalla collega di Paolo o fare dei mockup del funzionamento, metto a disposizione la mia conoscenza -relativa- di React.

Nel frattempo che viene creata la bozza grafica, domani rileggo quanto postato nel frattempo (ho avuto due giorni di fuoco) e torniamo a cercare di capire l'obiettivo del contenuto.

**Comment by SKApretto71** [comment]
Ciao,
l'ho usato anni fa, appena uscito, per un progetto in Inghilterra.
Chiedevo perché a occhio mi pare molto simile al servizio di AWS, ed il
livello di precisione nella trascrizione non pare essere male.
Concordo che è meglio usare quello che si conosce.

Non scrivo più codice da tempo, la mia era pura curiosità.
Purtroppo lato codice non potrò aiutare, proverò per la parte utente /
funzionalità.

Sulla parte "frasi celebri" direi che una volta risolto il tema
trascrizione + ricerca gioco diventerebbe una generalizzazionr
dell'approccio di ricerca di un nome di gioco composto da più parole
("Death Stranding", a caso, è come cercare "Mi fa stare bene"; se in
seguito la limitassimo ad un dizionario di brevi frasi). Così come avendo
la trascrizione ed una full text search potremmo usare un approccio per
ricerca semantica (e non più testuale) così da cercare argomenti e non
frasi / testi esatti. Entrambi partono dall'aver risolto il tema del nome
gioco, quindi mi concentrerei li.

Ah, sulla UI non chiedete a me, verrebbe fuori il pannello di controllo di
un driver per Windows 3.1 per la stampante ad aghi. :-)

Have fun!
---
Raffaele "SKApretto n.71" Sgherri

**Comment by SKApretto71** [comment]
Ciao,
il problema dell'opzione "B" è che va ripetuta per ogni
nuova/o punt-episodio, mentre la "A" una volta messa a punto ti salva
almeno dalla sgrossatura, puoi sempre avere una fase successiva di
revisione di qualità dove applicare "metadati" sulle parole chiave (quello
che al momento sono i titoli dei giochi) per discernere una ricorrenza a
valore da una citazione. Tecnicamente,per lo scopo iniziale ache le
citazioni sono da ricercare... altrimenti Death Stranding pesa come PacMan.
:-)

Più che una sbobinatura testuale, forse dovremmo pensare più ad un output
strutturato dove il flusso di parole viene analizzato a blocchi e se ha una
corrispondenza con le parole chiave annota l'offset audio. Possiamo quindi
tenerci il testuale per eventuali correzioni e per facilitare ricerche
documentali ma non ci sarebbe necessario per ricerche a campo fisso (solo i
giochi, il nostro MVP) e non libero.

---
Raffaele "SKApretto n.71" Sgherri

**Comment by pdonadeo** [comment]
Vengo da una giornata molto pesante, vi rispondo compiutamente domani. In breve però: Zampa mi ha mandato l'archivio delle scalette di tutti gli episodi e dopo averlo visto devo dire che è la svolta: sono completissime, c'è praticamente tutto. In prima battuta, ancora prima di sbobinare, trascrivere o usare un servizio, userei le scalette. Domani vi spiego meglio.

Per la grafica ho chiesto alla mia collega: nel fine settimana fa una bozza quindi lascerei stare la grafica e mi concentrerei sui contenuti.

**Reply by zampale** [reply]
Sono felice aiutino, anche se la così grande sorpresa nel fatto che le scalette siano complete un po' mi ha fatto male

**Reply by pdonadeo** [reply]
> Sono felice aiutino, anche se la così grande sorpresa nel fatto che le scalette siano complete un po' mi ha fatto male

La realtà è che non avevo la minima idea ci fosse così tanta preparazione: nel podcast siete molto "naturali" e sembrano chiacchiere a braccio.

**Reply by bionicco** [reply]
Da come parlavi delle scalette pensavo ci fosse quasi la trascrizione della puntata :D
Invece tutto sommato non c'è niente di più ne di meno della scaletta (che si sapeva esser fatta da un precisino)
Per alcuni episodi (dando un'occhiata a campione) poi non ci sono nemmeno tutti i titoli (non sempre ci sono i consigli e alcuni argomenti sono vaghi, vedi per esempio i consigli degli acquisti dell'ultimo episodio) e non ci salvano completamente dal ridare un'occhiata all'episodio.
Comunque sono d'accordo che ci forniscono una base solida con cui creare almeno una prima versione del DB

**Reply by pdonadeo** [reply]
Il punto non è che sulle scalette c'è tutto o non tutto ma c'è _abbastanza_ per mettere in piedi _qualcosa_.

Non averle avrebbe significato sbobinare o trascrivere in automatico e correggere con conseguente pubblicazione del motore di ricerca nel duemila e credici.

**Comment by SKApretto71** [comment]
Zampa for president. ;)
---
Raffaele Sgherri

**Comment by pdonadeo** [comment]
La collega sostiene che la pagina deve essere minimalista perché deve stare in poco spazio ed essere fruibile.

Le scalette di Zampa contengono, per ogni puntata, i giochi recensiti, quelli consigliati e per due delle tre stagioni anche una "cover" in formato PNG.

Quindi ho una proposta definitiva:
1. campo di ricerca stile Google
2. quando si inizia a digitare appare una lista di autocompletamento che pesca unicamente **tra i titoli dei giochi recensiti** e quindi niente immagini, proprio stile Google:
   ![image](https://user-images.githubusercontent.com/242049/167215459-8401e719-07c3-4197-9eaa-6c92b41bb0be.png)
4. una volta selezionata la voce (titolo) che interessa (touch su telefono, "freccia giù + invio" sul PC) appare la lista degli episodi in cui è stato citato il gioco.
5. ogni scheda consta di: titolo dell'episodio, la sua cover, link alla pagina del Post e minuti (circa) in cui se ne è parlato. Anche il minutaggio è presente in scaletta.

È minimale, niente ricerca full text per ora secondo me ma visto lo scarso tempo a disposizione è pure troppo. Io ho l'ambizione di andare on line con questa cosa prima della prossima uscita e, se buchiamo, certamente per quella dopo.

@MarcoJul per React perfetto, aspetto di avere la grafica e te la passo o la metto proprio nel repo.

Per le scalette... boh... io eviterei di includerle nel repository essendo materiale protetto da copyright. Se mi contattate personalmente vi passo lo ZIP con tutto.

Dobbiamo decidere un formato per il "database" che secondo me sarà poi un banale file di testo. JSON sarebbe ovviamente ultra comodo perché lo prendiamo così com'è ed è pronto da essere servito ma... personalmente detesto il JSON perché non è pensato per gli umani.

Idee? YAML? Boh...

**Reply by bionicco** [reply]
A me sembra perfetto come piano. Risponde assolutamente all'idea iniziale che avevo e il lavoro che c'è dietro è contenuto (al limite rifinibile in futuro con altre aggiunte).
I tempi desiderati mi sembrano un po' stretti, ma magari (ve lo auguro) avete più tempo libero di quanto ne abbia io ;)

Operativamente parlando se volete usare react non posso parteciapre attivamente allo sviluppo, ma magari posso comunque metterci mano per piccoli ritocchi. (con angular sarebbe un altro discorso)
Posso comunque occuparmi più della parte di data entry. Ora Paolo ti scrivo per avere le scalette, in modo da capire se elaborare direttamente i testo o se fare un'interfaccia per permettere a tutti di generare i contribudi al "DB"
A tal proposito io mi trovo molto a mio agio con json e mi sembra la scelta più immediata per un progetto JS, ma se preferisci qualcosa di più leggibile possiamo usare l'XML (ovviamente scherzo)

**Reply by pdonadeo** [reply]
_«Avevo un problema e ho pensato di usare XML. Ora ho due problemi»_

😂

**Comment by pdonadeo** [comment]
Dimenticavo: per contattarmi (per accesso al Drive delle scalette): p punto donadeo at gmail punto com

**Comment by SKApretto71** [comment]
Lato codice non posso aiutare, ma ad editare file (qualunque formato) per
trasferire i dati di scaletta volentierissimo.

Oramai sono piu da PowerPoint che da VS.Code ma il cervello funziona
ancora. ;)

Magari ci dividiamo il lavoro di data entry, il formato sceglietelo voi che
sapete. L'approccio minimale / MVP mi piace un sacco.
---
Raffaele Sgherri

**Reply by pdonadeo** [reply]
@SKApretto71 mandami in privato una email a p punto donadeo et gmail punto com che ti condivido la cartella di Drive.

**Comment by SKApretto71** [comment]
Te la rubo e la metto incorniciata in ufficio.

Raffaele Sgherri

**Comment by pdonadeo** [comment]
Non so se vi arrivano le email anche dei nuovi thread che scrivo qui su GitHub.

In caso non vi fosse arrivato nulla c'è una nuova novità: https://github.com/pdonadeo/e-uscito-joypad/discussions/10
